CommonTools/RecoAlgos/plugins/VertexCompositeCandidateCollectionSelector.cc

```diff
@@ -0,0 +1,174 @@
+  , pvNDOF_ ( iConfig.getParameter<int> ("pvNDOF") )
+  , label_  ( iConfig.getParameter<edm::InputTag>("v0").instance() )
+  , lxyCUT_      ( iConfig.getParameter<double>("lxyCUT") )
+  , lxyWRTbsCUT_ ( iConfig.getParameter<double>("lxyWRTbsCUT") )
+  , debug_ ( iConfig.getUntrackedParameter<bool>("debug") )
+{
+  if (debug_) std::cout << "VertexCompositeCandidateCollectionSelector::VertexCompositeCandidateCollectionSelector" << std::endl;
+  // product
+  produces<reco::VertexCompositeCandidateCollection>();
+
+  //now do what ever other initialization is needed
+  
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void
+VertexCompositeCandidateCollectionSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+   using namespace edm;
+   
+  if (debug_) std::cout << "VertexCompositeCandidateCollectionSelector::produce" << std::endl;
+
+   // Create auto_ptr for each collection to be stored in the Event
+   auto result = std::make_unique<reco::VertexCompositeCandidateCollection>();
+
+   edm::Handle<reco::BeamSpot> beamspotHandle;
+   iEvent.getByToken(bsToken_,beamspotHandle);
+   reco::BeamSpot const * bs = nullptr;
+   if (beamspotHandle.isValid())
+     bs = &(*beamspotHandle);
+
+
+   edm::Handle< reco::VertexCollection > pvHandle;
+   iEvent.getByToken(pvToken_, pvHandle );
+   reco::Vertex const * pv = nullptr;  
+   if (pvHandle.isValid()) {
+     pv = &pvHandle->front();
+     //--- pv fake (the pv collection should have size==1 and the pv==beam spot)
+     if (   pv->isFake() || pv->tracksSize()==0
+	    // definition of goodOfflinePrimaryVertex
+	    || pv->ndof() < pvNDOF_ || pv->z() > 24.)  pv = nullptr;
+   }
+
+   edm::Handle<reco::VertexCompositeCandidateCollection> v0Handle;
+   iEvent.getByToken(v0Token_, v0Handle);
+   int n = ( v0Handle.isValid() ? v0Handle->size() : -1 );
+   if (debug_) std::cout << "n: " << n << std::endl;
+   if (n>0) {
+  
+     reco::VertexCompositeCandidateCollection v0s = *v0Handle.product();    
+     for ( auto v0 : v0s ) {
+//       float mass     = v0.mass();
+//       float pt       = v0.pt();
+//       float p        = v0.p();
+//       float eta      = v0.eta();
+//       float phi      = v0.phi();
+//       int pdgID      = v0.pdgId();
+//       float chi2oNDF = v0.vertexNormalizedChi2();
+       GlobalPoint displacementFromPV = ( pv==nullptr ? dummyGP : GlobalPoint( (pv->x() - v0.vx()), 
+									       (pv->y() - v0.vy()), 
+									       0. ) );
+       GlobalPoint displacementFromBS = ( bs==nullptr ? dummyGP : GlobalPoint( -1*((bs->position().x() - v0.vx()) + (v0.vz() - bs->position().z()) * bs->dxdz()),
+									       -1*((bs->position().y() - v0.vy()) + (v0.vz() - bs->position().z()) * bs->dydz()), 
+									       0. ) );
+       float lxy      = ( pv==nullptr ? dummy : displacementFromPV.perp() );
+       float lxyWRTbs = ( bs==nullptr ? dummy : displacementFromBS.perp() );
```

Comment:
given that this selector implementation pretends to serve general VertexCompositeCandidate,
it would be better to either have appropriate variable naming or appropriate implementation.